File: dev/maya/plugin/ehm_plugins/ehm_replaceInfluence/replaceInfluence.cpp
/*
Author: Ehsan Hassani Moghaddam

Arguments and Flags:
first argument is the skinCluster name
-srcJnts (-s): list of joints that will have their weights removed from
-tgtJnt (-t): joint that weights from srcJnts go to
-errorOnMissingJnt (-e): When True, errors if tgt joint is not skinned to given skinCluster

Usage:
mc.loadPlugin("replaceInfluence", qt=True)

# this will give "a" and "b" weights to "c", even if "c" is not skinned to "box"
mc.replaceInfluence("box", s=["a", "b"], t="c")

# this will give "a" and "b" weights to "c", but errors if "c" is not skinned to "box"
mc.replaceInfluence("box", s=["a", "b"], t="c", e=True)

*/
#include "replaceInfluence.h"


MObject getSkinCluster(MDagPath& geo, MStatus& status)
{
    MObject node = geo.node();
    MItDependencyGraph itDG(node,
                            MFn::kSkinClusterFilter,
                            MItDependencyGraph::kUpstream);
    while (not itDG.isDone())
    {
        MObject oCurrentItem = itDG.currentItem();
        return oCurrentItem;
    }

    status = MStatus::kFailure;
    return MObject();
}

MDagPath dagFromName(MString& node, MStatus& status)
{
    MSelectionList mSel;
    mSel.add(node);
    MDagPath dag;
    status = mSel.getDagPath(0, dag);
    return dag;
}

bool containsElement(MDagPathArray array, MDagPath dag)
{
    for (unsigned int i=0; i<array.length(); ++i)
    {
        if (array[i] == dag)
        {
            return true;
        }
    }
    return false;
}

unsigned int getIndex(MDagPathArray array, MDagPath dag)
{
    for (unsigned int i=0; i<array.length(); ++i)
    {
        if (array[i] == dag)
        {
            return i;
        }
    }
    return -1;
}


MStatus ReplaceInfluenceClass::doIt(const MArgList& args)
{
    MStatus status = parseArguments(args);
    CHECK_MSTATUS_AND_RETURN_IT(status);
    return redoIt();
}

MStatus ReplaceInfluenceClass::redoIt()
{
    MStatus status = replaceInf();
    return status;
}

MStatus ReplaceInfluenceClass::replaceInf()
{
    MStatus status;

    // attach fn to skin
    geoShape.set(geo);
    geoShape.extendToShape();
    skin = getSkinCluster(geoShape, status);
    if (skin.isNull())
    {
        MGlobal::displayError( "\"" + geo.partialPathName() + "\" does not have a SkinCluster!");
        return status;
    }
    MFnSkinCluster fnSkin(skin);

    // list of all joints' dagPaths on skinCluster
    MDagPathArray infs;
    unsigned int numInfs = fnSkin.influenceObjects(infs);

    // index of tgtJnt on skinCluster
    unsigned int destIdx;

    // add tgtJnt to skinCluster if it's not already
    if (!containsElement(infs, tgtJnt))
    {        
        if (errorOnMissingTgt)
        {
            MGlobal::displayError("\"" + tgtJnt.partialPathName() + "\" is not skinned to \"" + geo.partialPathName() + "\"");
            return MStatus::kFailure;
        }
        else
        {
            MString cmd = "skinCluster -edit -ai \"" + tgtJnt.partialPathName() + "\" -lw 1 -wt 0 " + fnSkin.name() + ";";
            MGlobal::executeCommand(cmd);
            destIdx = numInfs;
            numInfs ++;
        }
    }    
    else
    {
        destIdx = getIndex(infs, tgtJnt);
    }

    // get list of IDs for verts and influences on skinCluster
    MFnMesh fnMesh(geoShape);
    uint numVtx = fnMesh.numVertices();
    
    MIntArray tmp1(numVtx);
    for(uint i=0; i<numVtx; ++i)
    {
        tmp1[i] = i;
    }

    MIntArray tmp2(numInfs);
    for(uint i=0; i<numInfs; ++i)
    {
        tmp2[i] = i;
    }
    
    vtxIds.copy(tmp1);
    infIds.copy(tmp2);
    
    // get old weights
    MFnSingleIndexedComponent fnComponents;
    components = fnComponents.create(MFn::kMeshVertComponent);
    MFnSingleIndexedComponent(components).addElements(vtxIds);
    MDoubleArray allWeights;
    status = fnSkin.getWeights(geoShape, components, infIds, allWeights);

    // 
    for (uint i=0; i<srcJnts.length(); ++i)
    {
        if (containsElement(infs, srcJnts[i]))
        {
            int srcIdx = getIndex(infs, srcJnts[i]);
            for (uint j=0; j<numVtx; ++j)
            {
                uint srcIndex = j * numInfs + srcIdx;
                uint destIndex = j * numInfs + destIdx;
                allWeights[destIndex] += allWeights[srcIndex];
                allWeights[srcIndex] = 0.0;
            }
        }
    }
    fnSkin.setWeights(geoShape, components, infIds, allWeights, false, &oldWeights);
    return status;
}

MStatus ReplaceInfluenceClass::undoIt()
{
    MStatus status;

    MFnSkinCluster fnSkin(skin);
    fnSkin.setWeights(geoShape, components, infIds, oldWeights, false, false);

    return status;
}

// CONSTRUCTOR:
ReplaceInfluenceClass::ReplaceInfluenceClass()
{
}

// DESTRUCTOR:
ReplaceInfluenceClass::~ReplaceInfluenceClass()
{
}

// FOR CREATING AN INSTANCE OF THIS COMMAND:
void* ReplaceInfluenceClass::creator()
{
   return new ReplaceInfluenceClass;
}

// MAKE THIS COMMAND UNDOABLE:
bool ReplaceInfluenceClass::isUndoable() const
{
   return true;
}

MStatus ReplaceInfluenceClass::parseArguments( const MArgList& args )
{
    MStatus status;
    
    MArgParser argData(syntax(), args);

    // get geo
    MString geoName = argData.commandArgumentString(0, &status);
    geo = dagFromName(geoName, status);
    // status = objExists(geo);
    if (status != MStatus::kSuccess)
    {
        MGlobal::displayError("Need a skinned object to replace joints' weights. Nothing was specified.");
        return status;
    }

    // get tgtJnt
    if (argData.isFlagSet("t"))
    {
        MString tgtJntName = argData.flagArgumentString("t", 0, &status);    
        tgtJnt = dagFromName(tgtJntName, status);
        // status = objExists(tgtJnt);
        if (status != MStatus::kSuccess)
        {
            MGlobal::displayError("tgtJnt (-t) flag expects a joint. Nothing was specified.");
            return status;
        }
    }
    else
    {
        MGlobal::displayError("tgtJnt (-t) flag expects a joint. But it was not set.");
        return MStatus::kFailure;
    }

    // get srcJnts
    if (argData.isFlagSet("s"))
    {
        unsigned int numOfSrcJnts = argData.numberOfFlagUses("s");
        for (unsigned int i=0; i<numOfSrcJnts; i++)
        {
            MArgList flagValue;
            status = argData.getFlagArgumentList("s", i, flagValue);
            MString srcJntName = flagValue.asString(0);
            MDagPath srcJnt = dagFromName(srcJntName, status);
            // status = objExists(flagValue.asString(0));
            if (status != MStatus::kSuccess)
            {
                MGlobal::displayError("srcJnts (-s) flag expects a list of joints. Nothing was specified.");
                return status;
            }            
            srcJnts.append(srcJnt);
        }
    }
    else
    {
        MGlobal::displayError("srcJnts (-s) flag expects a list of joints. But it was not set.");
        return MStatus::kFailure;
    }

    // get errorOnMissingTgt
    errorOnMissingTgt = false;
    if (argData.isFlagSet("e"))
    {
        errorOnMissingTgt = argData.flagArgumentBool("e", 0, &status);
    }

    return status;
}

MSyntax syntaxCreator()
{
    MSyntax syntax;
    syntax.addArg(MSyntax::kString);
    syntax.addFlag("-t", "-tgtJnt", MSyntax::kString);
    syntax.addFlag("-s", "-srcJnts", MSyntax::kString);
    syntax.addFlag("-e", "-errorOnMissingJnt", MSyntax::kBoolean);
    syntax.makeFlagMultiUse("-s");
    return syntax;
}

File: dev/scratch/oo_back/iRig/src/iRig/iRig_maya/plugin/scripted/replaceInfluence.py
"""
Author: Ehsan Hassani Moghaddam

Arguments and Flags:
first argument is the skinCluster name
-srcJnts (-s): list of joints that will have their weights removed from
-tgtJnt (-t): joint that weights from srcJnts go to
-errorOnMissingJnt (-e): When True, errors if tgt joint is not skinned to given skinCluster


Usage:
mc.loadPlugin('replaceInfluence', qt=True)

# this will give 'a' and 'b' weights to 'c', even if 'c' is not skinned to 'box'
mc.replaceInfluence('box', s=['a', 'b'], t='c')

# this will give 'a' and 'b' weights to 'c', but errors if 'c' is not skinned to 'box'
mc.replaceInfluence('box', s=['a', 'b'], t='c', e=True)


"""


import sys
import maya.cmds as mc
import maya.api.OpenMaya as om2
import maya.api.OpenMayaAnim as oma2


class ReplaceInfluenceClass(om2.MPxCommand):
    kPluginCmdName = 'replaceInfluence'


    def __init__(self):
        om2.MPxCommand.__init__(self)
        self.geo = None
        self.srcJnts = None
        self.tgtJnt = None
        self.errorOnMissingTgt = False

    @staticmethod
    def cmdCreator():
        return ReplaceInfluenceClass()

    def doIt(self, args):
        self.parseArguments(args)
        self.redoIt()

    def redoIt(self):
        self.replaceInf()


    def undoIt(self):
        self.fnSkin.setWeights(self.dagShape, self.components, self.infIds, self.oldWeights, False, False)

    def isUndoable(self):
        return True

    def parseArguments(self, args):
        argData = om2.MArgParser(self.syntax(), args)

        self.geo = argData.commandArgumentString(0)

        if argData.isFlagSet('t'):
            self.tgtJnt = argData.flagArgumentString('t', 0)

        self.srcJnts = []
        if argData.isFlagSet('s'):
            for i in range(argData.numberOfFlagUses('s')):
                flagValue = argData.getFlagArgumentList('s', i)
                self.srcJnts.append(flagValue.asString(0))

        if argData.isFlagSet('e'):
            self.errorOnMissingTgt = argData.flagArgumentBool('e', 0)

        for node in self.srcJnts + [self.geo, self.tgtJnt]:
            if not mc.objExists(node):
                mc.error('Object not found: "{}"'.format(node))

    def replaceInf(self):
        self.fnSkin = self.getFnSkinCluster(self.geo)
        self.dagShape = self.getDagShape(self.geo)
        infs = self.fnSkin.influenceObjects()
        numInfs = len(infs)
        infNames = map(lambda x: x.partialPathName(), infs)

        if self.tgtJnt not in infNames:
            if self.errorOnMissingTgt:
                raise RuntimeError('"{}" is not skinned to "{}"'.format(self.tgtJnt, self.geo))
            else:
                mc.skinCluster(self.fnSkin.name(), edit=True, ai=self.tgtJnt, lw=True, wt=0)
                destIdx = len(infNames)
                numInfs += 1
        else:
            destIdx = infNames.index(self.tgtJnt)

        numVtx = om2.MFnMesh(self.dagShape).numVertices


        vtxIds = om2.MIntArray(range(numVtx))
        self.infIds = om2.MIntArray(range(numInfs))

        self.components = om2.MFnSingleIndexedComponent().create(om2.MFn.kMeshVertComponent)
        om2.MFnSingleIndexedComponent(self.components).addElements(vtxIds)

        allWeights = self.fnSkin.getWeights(self.dagShape, self.components, self.infIds)
        self.oldWeights = allWeights[:]


        for s in self.srcJnts:
            try:
                srcIdx = infNames.index(s)
            except ValueError:
                continue
            for i in vtxIds:
                srcIndex = i * numInfs + srcIdx
                destIndex = i * numInfs + destIdx
                allWeights[destIndex] += allWeights[srcIndex]
                allWeights[srcIndex] = 0.0
        self.fnSkin.setWeights(self.dagShape, self.components, self.infIds, allWeights, False, False)

    def getFnSkinCluster(self, geo):
        # if name of skinCluster is given
        mSel = om2.MSelectionList()
        mSel.add(geo)
        depN = mSel.getDependNode(0)
        if depN.hasFn(om2.MFn.kSkinClusterFilter):
            return oma2.MFnSkinCluster(depN)

        # if name of geo is given
        dagShape = self.getDagShape(geo)
        try:
            itDG = om2.MItDependencyGraph(dagShape.node(),
                                          om2.MFn.kSkinClusterFilter,
                                          om2.MItDependencyGraph.kUpstream)
            while not itDG.isDone():
                oCurrentItem = itDG.currentNode()
                return oma2.MFnSkinCluster(oCurrentItem)
        except:
            pass
        return None

    def getDagShape(self, geo):
        mSel = om2.MSelectionList()
        mSel.add(geo)
        dagS = mSel.getDagPath(0)
        dagS.extendToShape()
        return dagS


def syntaxCreator():
    syntax = om2.MSyntax()
    syntax.addArg(om2.MSyntax.kString)
    syntax.addFlag('-t', '-tgtJnt', om2.MSyntax.kString)
    syntax.addFlag('-s', '-srcJnts', om2.MSyntax.kString)
    syntax.addFlag('-e', '-errorOnMissingJnt', om2.MSyntax.kBoolean)
    syntax.makeFlagMultiUse('-s')
    return syntax


def maya_useNewAPI():
    pass


def initializePlugin(mobject):
    mplugin = om2.MFnPlugin(mobject)
    try:
        mplugin.registerCommand(ReplaceInfluenceClass.kPluginCmdName,
                                ReplaceInfluenceClass.cmdCreator,
                                syntaxCreator)
    except:
        sys.stderr.write('Failed to register command: ' + ReplaceInfluenceClass.kPluginCmdName)


def uninitializePlugin(mobject):
    mplugin = om2.MFnPlugin(mobject)
    try:
        mplugin.deregisterCommand(ReplaceInfluenceClass.kPluginCmdName)
    except:
        sys.stderr.write('Failed to unregister command: ' + ReplaceInfluenceClass.kPluginCmdName)


